tests/std/tests/P2136R3_invoke_r/test.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <functional>
+#include <type_traits>
+
+using namespace std;
+
+// TRANSITION, DevCom-1457457
+namespace detail {
+    static constexpr bool permissive() {
+        return false;
+    }
+
+    template <class>
+    struct DependentBase {
+        static constexpr bool permissive() {
+            return true;
+        }
+    };
+
+    template <class T>
+    struct Derived : DependentBase<T> {
+        static constexpr bool test() {
+            return permissive();
+        }
+    };
+} // namespace detail
+constexpr bool is_permissive = detail::Derived<int>::test();
+
+constexpr int square(int n) {
+    return n * n;
+}
+constexpr int square_noexcept(int n) noexcept {
+    return n * n;
+}
+
+struct Thing {
+    int n = 0;
+    constexpr int& foo() {
+        return n;
+    }
+};
+
+constexpr bool is_rvalue(const int&) {
+    return false;
+}
+constexpr bool is_rvalue(int&&) {
+    return true;
+}
+
+constexpr bool test_invoke_r() {
+    auto v1 = invoke_r<long int>(square, 3);
+    assert(v1 == 9L);
+    static_assert(is_same_v<decltype(v1), long int>);
+
+    auto v2 = invoke_r<double>([]() -> int { return 5; });
+    assert(v2 == 5);
+    static_assert(is_same_v<decltype(v2), double>);
+    static_assert(is_void_v<decltype(invoke_r<void>(square, 1))>);
+
+    // TRANSITION, DevCom-1457457
+    static_assert(noexcept(invoke_r<int>(square, 3)) == is_permissive, "invoke_r<int>(square, 3) is noexcept");
+    static_assert(noexcept(invoke(square, 3)) == is_permissive, "invoke(square, 3) is noexcept");
+#ifdef __cpp_noexcept_function_type
+    static_assert(noexcept(invoke_r<int>(square_noexcept, 3)), "invoke_r<int>(square_noexcept, 3) isn't noexcept");
+    static_assert(noexcept(invoke(square_noexcept, 3)), "invoke(square_noexcept, 3) isn't noexcept");
+#endif

```

Comment:
This should probably be commented according to our current (albeit not-totally-rigorous) convention.

Reply:
I thought a couple lines was an exception?

Reply:
I tend to believe that the only numbers that exist are 0, 1, and infinity, but I guess 2 is fine here.

Reply:
Now it's 4 after recent changes, though! :scream_cat:

Reply:
It's now within a number that exists 🎉.